api/MarbleCollectorApi/Controllers/AssignmentsController.cs

```diff
@@ -1,60 +1,139 @@
+using MarbleCollectorApi.Data.Repository;
 using MarbleCollectorApi.ViewModels;
+using Microsoft.EntityFrameworkCore.ChangeTracking;
+using Microsoft.AspNetCore.SignalR;
+using MarbleCollectorApi.Hubs;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Authorization;
 
 namespace MarbleCollectorApi.Controllers
 {
+    [Authorize]
     [Route("api/[controller]")]
     [ApiController]
+    [ProducesResponseType(StatusCodes.Status401Unauthorized)]
     public class AssignmentsController : Controller
     {
+        private readonly IAssignmentRepository _assignmentRepository;
+        private readonly IHubContext<ParentNotificationHub> _parentNotificationHubContext;
+        private readonly IHubContext<ChildrenNotificationHub> _childrenNotificationHubContext;
+
+        public AssignmentsController(IAssignmentRepository assignmentRepository, IHubContext<ParentNotificationHub> parentNotificationHubContext, IHubContext<ChildrenNotificationHub> childrenNotificationHubContext)
+        {
+            _assignmentRepository = assignmentRepository ?? throw new ArgumentNullException(nameof(assignmentRepository));
+            _parentNotificationHubContext = parentNotificationHubContext;
+            _childrenNotificationHubContext = childrenNotificationHubContext;
+        }
+
         // TODO js (04.03.2021): Can all users get all assignments?
         [HttpGet]
         [ProducesResponseType(StatusCodes.Status200OK)]
         public ActionResult<IEnumerable<Assignment>> GetAssignments()
         {
-            return NotFound(); // TODO js (04.03.2021): To be implemented!
+            return Ok(_assignmentRepository.GetAll().Select(assignment => assignment.Map()));
         }
 
         // TODO js (04.03.2021): Can all users get all assignments?
         [HttpGet("{id}")]
         [ProducesResponseType(StatusCodes.Status200OK)]
         public ActionResult<Assignment> GetAssignmentById(int id)
         {
-            return NotFound(); // TODO js (04.03.2021): To be implemented!
+            var assignment = _assignmentRepository.GetSingle(id);
+            if (assignment == null)
+            {
+                return NotFound();
+            }
+
+            return Ok(assignment.Map());
         }
 
         [HttpPost()]
+        [Authorize(Roles = Const.UserRoleParent)]
         [ProducesResponseType(StatusCodes.Status201Created)]
         [ProducesResponseType(StatusCodes.Status400BadRequest)]
+        [ProducesResponseType(StatusCodes.Status403Forbidden)]
         public ActionResult<Assignment> CreateAssignment(Assignment assignment)
         {
-            return NotFound(); // TODO js (04.03.2021): To be implemented!
+            var entityEntry = _assignmentRepository.Add(assignment.Map());
+
+            try
+            {
+                _assignmentRepository.Commit();
+            }
+            catch
+            {
+                // TODO hs 210307, is i.e. foreign key constraint a bad request or internal server error?
```

Comment:
@StuderS41 Das könnte auch einen **403** zurückliefern (aus Wikipedia):
**403 Forbidden**
The request contained valid data and was understood by the server, but the server is refusing action. This may be due to the user not having the necessary permissions for a resource or needing an account of some sort, **or attempting a prohibited action (e.g. creating a duplicate record where only one is allowed)**. This code is also typically used if the request provided authentication by answering the WWW-Authenticate header field challenge, but the server did not accept that authentication. The request should not be repeated.